File: app/src/main/java/com/kelsos/mbrc/commands/model/UpdatePlaylistList.java
package com.kelsos.mbrc.commands.model;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.google.inject.Inject;
import com.kelsos.mbrc.data.Playlist;
import com.kelsos.mbrc.interfaces.ICommand;
import com.kelsos.mbrc.interfaces.IEvent;
import com.kelsos.mbrc.model.MainDataModel;
import rx.Observable;
import rx.Subscriber;
import rx.schedulers.Schedulers;
import timber.log.Timber;

public class UpdatePlaylistList implements ICommand {

  private final MainDataModel model;
  private ObjectMapper mapper;

  @Inject public UpdatePlaylistList(MainDataModel model, ObjectMapper mapper) {
    this.model = model;
    this.mapper = mapper;
  }

  @Override public void execute(IEvent e) {
    ArrayNode nodes = (ArrayNode) e.getData();
    Observable.from(nodes).flatMap(node -> Observable.create((Subscriber<? super Playlist> subscriber) -> {
      try {
        Playlist playlist = mapper.treeToValue(node, Playlist.class);
        subscriber.onNext(playlist);
        subscriber.onCompleted();
      } catch (JsonProcessingException e1) {
        subscriber.onError(e1);
      }
    })).subscribeOn(Schedulers.io()).toList().subscribe(model::setPlaylists, t -> {
      Timber.v(t, "failed to parse the playlists");
    });
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/commands/model/UpdatePlaylistList.kt
package com.kelsos.mbrc.commands.model

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ArrayNode
import com.kelsos.mbrc.data.Playlist
import com.kelsos.mbrc.interfaces.ICommand
import com.kelsos.mbrc.interfaces.IEvent
import com.kelsos.mbrc.model.MainDataModel
import rx.AsyncEmitter
import rx.Observable
import rx.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Inject

class UpdatePlaylistList
@Inject constructor(private val model: MainDataModel, private val mapper: ObjectMapper) : ICommand {

  override fun execute(e: IEvent) {
    val nodes = e.data as ArrayNode
    Observable.from(nodes).flatMap {
      Observable.fromEmitter({ it: AsyncEmitter<Playlist> ->
        try {
          val playlist = mapper.treeToValue<Playlist>(nodes, Playlist::class.java)
          it.onNext(playlist)
          it.onCompleted()
        } catch (e1: JsonProcessingException) {
          it.onError(e1)
        }
      }, AsyncEmitter.BackpressureMode.BUFFER)
    }.subscribeOn(Schedulers.io()).toList().subscribe({ model.setPlaylists(it) }) {
      Timber.v(it, "failed to parse the playlists")
    }
  }
}
